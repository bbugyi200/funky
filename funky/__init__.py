"""
Funky takes shell functions to the next level by making them easier to define,
more flexible, and more interactive.
"""

__author__ = """Bryan M Bugyi"""
__email__ = "bryanbugyi34@gmail.com"
__version__ = "3.5.3"
