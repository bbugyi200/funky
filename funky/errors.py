"""This module holds all custom exception class definitions."""

from __future__ import absolute_import, division, print_function

from builtins import (  # noqa: F401  # pylint: disable=redefined-builtin,unused-import
    bytes,
    dict,
    int,
    range,
    str,
    super,
)
from typing import Any


class FunkyError(Exception):
    """Base custom exception class."""

    def __init__(self, *args: Any, **kwargs: Any) -> None:
        returncode: int = kwargs.pop("returncode", 1)

        super().__init__(*args, **kwargs)  # type: ignore
        self.returncode = returncode


class FunkNotDefinedError(FunkyError):
    """
    Raised when an undefined funk is referenced in a mannor that is not allowed.
    """

    def __init__(self, *args: Any, **kwargs: Any) -> None:
        funk = kwargs.pop("funk", None)
        global_ = kwargs.pop("global_", False)

        if funk is None:
            if global_:
                msg = "No global funks are defined."
            else:
                msg = "No local funks are defined in the current directory."
        else:
            msg_fmt = (
                '"{}" does not match any local funks defined in the current '
                "directory.".format(funk)
            )
            msg = msg_fmt.format(funk)

        super().__init__(msg, *args, **kwargs)


class ArgumentError(FunkyError):
    """Raised when the given command-line arguments fail validation check."""


class BlankDefinition(FunkyError):
    """
    Raised when the user attempts to define a funk using a blank definition.
    """
