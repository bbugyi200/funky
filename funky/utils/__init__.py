"""Miscellaneous utility functions."""

import sys
import termios
import tty


def getch(prompt: str = None) -> str:
    """Reads a single character from stdin.

    Args:
        prompt (optional): prompt that is presented to user.

    Returns:
        The single character that was read.
    """
    if prompt:
        sys.stdout.write(prompt)

    sys.stdout.flush()

    fd = sys.stdin.fileno()
    old_settings = termios.tcgetattr(fd)
    try:
        tty.setraw(sys.stdin.fileno())
        ch = sys.stdin.read(1)
    finally:
        termios.tcsetattr(fd, termios.TCSADRAIN, old_settings)
    return ch
