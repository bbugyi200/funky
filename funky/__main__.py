"""Entry point for funky. Simply execute funky (i.e. run `python funky`)."""

import importlib
import os
import sys


if os.environ.get("DEBUG_FUNKY", None):
    parent_dir = os.path.abspath(os.path.dirname(__file__))
    sys.path.append(os.path.dirname(parent_dir))
    __package__ = os.path.basename(  # pylint: disable=redefined-builtin
        parent_dir
    )
    importlib.import_module(__package__)

from funky import app


if __name__ == "__main__":
    sys.exit(app.main())
