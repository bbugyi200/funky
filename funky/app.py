"""Initialize and run funky.

Do not execute this file directly. Use __main__.py by executing this modules
containing directory.  In other words, use `python funky`.
"""

import argparse
from importlib.resources import read_text
from pathlib import Path
import sys
from typing import List, Optional, Sequence, Union

import funky
from funky import commands, errors
from funky.utils import log


_SUPPORTED_SHELLS = ["bash", "zsh"]


def main(argv: Sequence[str] = None) -> int:
    try:
        if argv is None:
            argv = sys.argv[1:]

        verbose = any(x in argv for x in ["--verbose", "-v", "-hv", "-vh"])
        parser = _get_argparser(verbose=verbose)
        args = parser.parse_args(argv)

        log.init_logger(debug=args.debug, verbose=args.verbose)
        log.logger.debug("Starting funky.")
        log.logger.vdebug("argv = {}".format(argv))  # type: ignore
        log.logger.vdebug("Command-line Arguments: {}".format(args))  # type: ignore

        if args.init:
            return run_init(args.init)

        if args.setup_shell:
            return run_setup_shell(args.setup_shell)

        _CmdAction.command(args)
    except errors.ArgumentError as e:
        log.logger.error("%s\n", str(e))
        parser.print_usage()
        return 2
    except errors.FunkyError as e:
        log.logger.debug("Exit Status: %s", e.returncode)
        log.logger.error(str(e))
        return e.returncode
    except Exception as e:
        log.logger.exception("%s: %s", type(e).__name__, str(e))
        raise

    return 0


def run_init(shell: str) -> int:
    del shell  # all supported shells use the same code ATM

    funky_sh = read_text("scripts.shell", "funky.sh")
    print(funky_sh, end="")
    return 0


def run_setup_shell(shell: str) -> int:
    if shell == "zsh":
        config_base = ".zshrc"
    else:
        assert shell == "bash"
        config_base = ".bashrc"

    config_path = Path.home() / config_base
    funky_init = f"funky --init {shell}"
    if not config_path.exists() or funky_init not in config_path.read_text():
        with config_path.open("a") as f:
            cmd = (
                "\n# setup funky\n"
                "command -v funky &>/dev/null &&"
                f' eval "$({funky_init})"'
            )
            log.logger.info(
                "Appending %d lines to your %s file...",
                len(cmd.split("\n")),
                config_base,
            )
            f.write(cmd)
    return 0


def _get_argparser(verbose: bool = False) -> argparse.ArgumentParser:
    """Get command-line arguments.

    Args:
        verbose (bool): If True, do not suppress the help output of any
                        arguments.

    Returns:
        argparse.ArgumentParser object.
    """
    parser = argparse.ArgumentParser(prog="funky", description=funky.__doc__)

    init_group = parser.add_mutually_exclusive_group()
    init_group.add_argument(
        "--setup-shell",
        choices=_SUPPORTED_SHELLS,
        help="Ensure your shell is configured correctly.",
    )
    init_group.add_argument(
        "--init",
        choices=_SUPPORTED_SHELLS,
        help=(
            "Initialize your shell environments. This should be run from your"
            " .bashrc / .zshrc file."
        ),
    )

    parser.add_argument(
        "-d", "--debug", action="store_true", help="Enable debug mode."
    )
    parser.add_argument(
        "-v", "--verbose", action="store_true", help="Enable verbose output."
    )
    parser.add_argument(
        "--version",
        action="version",
        version="%(prog)s {}".format(funky.__version__),
    )
    parser.add_argument(
        "--color",
        nargs=1,
        choices=("y", "n"),
        default="n",
        help="Colorize funk definitions.",
    )

    parser.add_argument(
        "-g",
        "--global",
        dest="global_",
        action="store_true",
        help=("Enable global scope." if verbose else argparse.SUPPRESS),
    )

    def format_docstring(doc: Optional[str]) -> str:
        """Converts command docstring to argparse help doc"""
        if doc is None:
            doc = ""
        return doc.strip().replace("\n", " ")

    command_group = parser.add_argument_group(
        title="Action Commands",
        description=(
            "All of these options act on the current set of local funks in "
            "some way. If no action command is provided, the default action "
            "is to display all of the local funks currently in scope. These "
            "commands are mutually exclusive."
        ),
    )
    command_group.add_argument(
        _CmdFlag.ADD,
        nargs=1,
        dest="command_args",
        action=_CmdAction,
        metavar="FUNK",
        help=format_docstring(commands.Add.__doc__),
    )
    command_group.add_argument(
        _CmdFlag.REMOVE,
        nargs="?",
        dest="command_args",
        action=_CmdAction,
        metavar="FUNK",
        help=format_docstring(commands.Remove.__doc__),
    )
    command_group.add_argument(
        _CmdFlag.EDIT,
        nargs=1,
        dest="command_args",
        action=_CmdAction,
        metavar="FUNK",
        help=format_docstring(commands.Edit.__doc__),
    )
    command_group.add_argument(
        _CmdFlag.RENAME,
        nargs=2,
        dest="command_args",
        action=_CmdAction,
        metavar=("OLD", "NEW"),
        help=format_docstring(commands.Rename.__doc__),
    )
    command_group.add_argument(
        "command_args",
        nargs="?",
        action=_CmdAction,
        metavar="FUNK",
        help=format_docstring(commands.Show.__doc__),
    )

    return parser


class _CmdAction(argparse.Action):
    """Custom ArgumentParser Action for Action Commands"""

    flag: Optional[str] = None
    option_string = None

    def __call__(
        self,
        parser: argparse.ArgumentParser,
        namespace: argparse.Namespace,
        values: Union[None, str, Sequence[str]],
        option_string: str = None,
    ) -> None:
        if self.__class__.flag is None:
            self.__class__.flag = option_string
            self.__class__.option_string = option_string
        elif option_string is not None:
            raise errors.ArgumentError(
                "Option {} can not be used with option {}. All action "
                "commands are mutually exclusive.".format(
                    option_string, self.__class__.option_string
                )
            )
        else:
            return

        dest: List[Optional[str]]
        if values is None or isinstance(values, str):
            dest = [values]
        else:
            dest = list(values)

        setattr(namespace, self.dest, dest)

    @classmethod
    def command(cls, args: argparse.Namespace) -> None:
        """Map from actions to commands."""
        cmd_builder = {
            _CmdFlag.ADD: commands.Add,
            _CmdFlag.REMOVE: commands.Remove,
            _CmdFlag.EDIT: commands.Edit,
            _CmdFlag.RENAME: commands.Rename,
            _CmdFlag.SHOW: commands.Show,
        }[cls.flag]

        cmd = cmd_builder(  # type: ignore[abstract]
            args.command_args,
            color=(args.color[0] == "y"),
            global_=args.global_,
            verbose=args.verbose,
        )
        return cmd()


class _CmdFlag:
    """Command Flags

    The value of each command flag will be used by argparse to generate the
    option string corresponding to that command.
    """

    ADD = "-a"
    REMOVE = "-r"
    EDIT = "-e"
    SHOW = None
    RENAME = "-R"
