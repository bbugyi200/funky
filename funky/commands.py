"""Command definitions."""

from __future__ import absolute_import, division, print_function

from abc import ABCMeta, abstractmethod
from builtins import (  # noqa: F401  # pylint: disable=redefined-builtin,unused-import
    bytes,
    dict,
    int,
    range,
    str,
    super,
)
import json
import os
import re
import shlex
import subprocess as sp
import tempfile
from typing import Dict, List, Union

from pygments import highlight  # type: ignore
from pygments.formatters import TerminalFormatter  # type: ignore
from pygments.lexers import BashLexer  # type: ignore

from funky import errors, utils
from funky.utils import log


class Command:
    """Abstract base command class.

    To use a command, the corresponding command class should be used to build a
    command instance.  A command instance is a callable object.

    Args:
        args (iter): The first element is necessarily a funk name. The others
                     elements (if any exist) vary depending on what command is
                     being used.
        color (bool): If True, colorize output (if command produces output).
        global_ (bool): If True, the global database will be used instead of the
                        local database.
        verbose (bool): If True, Show command displays verbose command
                        definition.

    IMPORTANT: The class docstring of a Command subclass is used by argparse to
               generate output for the help command.
    """

    __metaclass__ = ABCMeta

    FUNKY_DB_FILENAME = ".funky"
    GLOBAL_FUNKY_DB_FILENAME = "{}/.funky".format(os.path.expanduser("~"))

    def __init__(
        self,
        args: Union[None, str, List[str]],
        color: bool = False,
        global_: bool = False,
        verbose: bool = False,
    ) -> None:
        if isinstance(args, str):
            args = [args]

        if global_:
            self.ACTIVE_DB_FILENAME = self.GLOBAL_FUNKY_DB_FILENAME
        else:
            self.ACTIVE_DB_FILENAME = self.FUNKY_DB_FILENAME

        if args is None:
            self.funk = None
            self.args = []
        else:
            self.funk = args[0]
            self.args = args[1:]

        self.color = color
        self.verbose = verbose
        self.global_ = global_
        self.funk_dict = self.load(self.ACTIVE_DB_FILENAME)

        log.logger.vdebug("Existing Funks: {}".format(self.funk_dict))  # type: ignore

    @staticmethod
    def abort() -> None:
        """Print abort message."""
        print()
        log.logger.info("OK. Aborting...")

    def purge_db(self) -> None:
        """Removes the database file."""
        try:
            os.remove(self.ACTIVE_DB_FILENAME)
        except (IOError, OSError):
            pass

    def commit(self) -> None:
        """Saves funk changes to database."""
        if self.funk_dict:
            log.logger.debug(
                "Committing changes to database: %s", self.ACTIVE_DB_FILENAME
            )
            with open(self.ACTIVE_DB_FILENAME, "w") as f:
                json.dump(self.funk_dict, f)
        else:
            log.logger.debug("Removing %s.", self.ACTIVE_DB_FILENAME)
            self.purge_db()

    @staticmethod
    def load(DB_FILENAME: str) -> Dict[str, str]:
        try:
            with open(DB_FILENAME, "r") as f:
                result: Dict[str, str] = json.load(f)
                return result
        except (IOError, OSError):
            return {}

    @abstractmethod
    def __call__(self) -> None:
        log.logger.debug("Running %s command.", self.__class__.__name__)


class Show(Command):
    """
    When no action command is specified, the default action is to display
    existing funks. An funk name (FUNK) can optionally be provided as an
    argument to display only FUNK. If FUNK ends in two periods ('..'), it is
    treated as a prefix instead of an exact match: all funks that start with
    FUNK (not including the trailing '..') will be displayed.
    """

    def show(self, funk: str) -> None:
        """Print funk and funk command definition to stdout."""
        cmd_string = self.funk_dict[funk]
        if "\n" in cmd_string:
            show_output = "{0}() {{\n\t{1}\n}}".format(
                funk, cmd_string.replace("\n", "\n\t")
            )
        else:
            show_output = "{0}() {{ {1}; }}".format(funk, cmd_string)

        if self.verbose:
            unalias_out = "unalias {} &> /dev/null\n".format(funk)

            show_output = "".join([unalias_out, show_output])

        if self.color:
            final_output = highlight(
                show_output, BashLexer(), TerminalFormatter()
            ).strip()
        else:
            final_output = show_output

        print(final_output)

    def show_search(self, prefix: str) -> None:
        """Prints all funks that start with @prefix to stdout."""
        log.logger.debug("Running show command for all defined funks.")
        sorted_funks = sorted(
            [funk for funk in self.funk_dict if funk.startswith(prefix)],
            key=lambda x: x.lower(),
        )

        if not sorted_funks:
            raise errors.FunkNotDefinedError(funk=self.funk)

        for funk in sorted_funks:
            if self.verbose:
                print()
            self.show(funk)

    def __call__(self) -> None:
        super().__call__()
        if not self.funk_dict:
            self.purge_db()
            raise errors.FunkNotDefinedError(global_=self.global_)

        if self.funk is None:
            self.show_search(prefix="")
        elif self.funk[-2:] == "..":
            self.show_search(prefix=self.funk[:-2])
        elif self.funk not in self.funk_dict:
            raise errors.FunkNotDefinedError(funk=self.funk)
        else:
            self.show(self.funk)


class Rename(Command):
    """Rename an existing funk. OLD funk is renamed to NEW."""

    def __call__(self) -> None:
        super().__call__()
        if self.funk not in self.funk_dict:
            raise errors.FunkNotDefinedError(funk=self.funk)

        new_funk = self.args[0]
        if new_funk in self.funk_dict:
            y_or_n = utils.getch(
                '"{}" is already in use. Overwrite? (y/n): '.format(new_funk)
            )
            if y_or_n == "y":
                print()
            else:
                return self.abort()

        self.funk_dict[new_funk] = self.funk_dict[self.funk]
        self.funk_dict.pop(self.funk)

        log.logger.info(
            'Funk "%s" has successfully been renamed to "%s".',
            self.funk,
            new_funk,
        )
        self.commit()


class Edit(Command):
    """Edit an existing funk."""

    NONE_FUNK_ERROR = (
        "Logic Error: Must provide a funk name when using the --edit option."
    )

    def remove_funk(self) -> None:
        """Removes the funk defined at instance creation time."""
        assert self.funk is not None, self.NONE_FUNK_ERROR
        self.funk_dict.pop(self.funk)
        log.logger.info('Removed funk "%s".', self.funk)

    def edit_funk(self, startinsert: bool = False) -> None:
        """Opens up funk definition using temp file in $EDITOR for editing.

        Args:
            funk (optional): The funk to edit. If not given, this function uses
                             the funk defined at instance creation time.

        Returns (str):
            Contents of temp file after $EDITOR closes.
        """
        tf = tempfile.NamedTemporaryFile(
            prefix="{}.".format(self.funk),
            suffix=".sh",
            dir="/var/tmp",
            mode="w",
            delete=False,
        )
        if self.funk in self.funk_dict:
            tf.write(self.funk_dict[self.funk])
        tf.close()

        editor_cmd_list = self._editor_cmd_list(startinsert)

        editor_cmd_list.append(tf.name)
        try:
            sp.check_call(editor_cmd_list)
        except sp.CalledProcessError as e:
            raise errors.FunkyError(
                "Failed to open editor using: {}".format(editor_cmd_list)
            ) from e

        tf = open(tf.name, "r")
        edited_cmd_string = tf.read()
        tf.close()
        os.unlink(tf.name)

        if edited_cmd_string.strip() == "":
            raise errors.BlankDefinition("Funk definition cannot be blank.")

        log.logger.debug('New Command String: "%s"', edited_cmd_string)
        formatted_cmd_string = self._apply_shortcuts(edited_cmd_string.strip())

        assert self.funk is not None, self.NONE_FUNK_ERROR
        self.funk_dict[self.funk] = formatted_cmd_string

    @staticmethod
    def _editor_cmd_list(startinsert: bool = False) -> List[str]:
        """Generates and returns editor command list."""
        if "EDITOR" in os.environ:
            editor_cmd_list = shlex.split(os.environ["EDITOR"])
            log.logger.debug(
                "Editor command set to $EDITOR: %s", editor_cmd_list
            )
        else:
            editor_cmd_list = ["vim"]
            log.logger.debug(
                "Editor command falling back to default: %s",
                editor_cmd_list,
            )

        if any("vim" in arg for arg in editor_cmd_list) and startinsert:
            editor_cmd_list.append("+startinsert")

        return editor_cmd_list

    @staticmethod
    def _apply_shortcuts(cmd_string: str) -> str:
        """Formats command string for correct execution and display."""
        if cmd_string.startswith("@./"):
            cmd_string = 'cd {}/"$@" || return 1'.format(
                cmd_string.replace("@./", "{}/".format(os.getcwd()))
            )

        double_quoted_conds = [
            cmd_string.startswith('"'),
            cmd_string.endswith('"'),
        ]
        single_quoted_conds = [
            cmd_string.startswith("'"),
            cmd_string.endswith("'"),
        ]
        if all(double_quoted_conds) or all(single_quoted_conds):
            cmd_string = "echo {}".format(cmd_string)

        bad_keys = [r"\$", r"\n", "return", "done", "fi"]
        if re.search("({})".format("|".join(bad_keys)), cmd_string):
            new_cmd_string = cmd_string
        else:
            new_cmd_string = '{} "$@"'.format(cmd_string)

        return new_cmd_string

    def __call__(self) -> None:
        super().__call__()
        if self.funk and self.funk not in self.funk_dict:
            raise errors.FunkNotDefinedError(funk=self.funk)

        try:
            self.edit_funk()
            log.logger.info('Edited funk "%s".', self.funk)
        except errors.BlankDefinition as e:
            log.logger.info(str(e))
            self.remove_funk()

        self.commit()


class Remove(Edit):
    """
    Remove an existing funk. Or (if FUNK is not given) remove all funks defined
    in this directory.
    """

    def __call__(self) -> None:
        Command.__call__(self)
        if self.funk and self.funk not in self.funk_dict:
            raise errors.FunkNotDefinedError(funk=self.funk)

        if not self.funk_dict:
            raise errors.FunkNotDefinedError(global_=self.global_)

        if self.funk is None:
            log.logger.debug("Prompting to destroy local funk database.")
            prompt = (
                "Remove all local funks defined in this directory? (y/n): "
            )
            y_or_n = utils.getch(prompt)
            if y_or_n == "y":
                self.funk_dict = {}
                print()
                log.logger.info(
                    "Done. The local funk database has been removed."
                )
            else:
                return self.abort()
        else:
            self.remove_funk()

        self.commit()


class Add(Edit):
    """Add a new funk."""

    def __call__(self) -> None:
        Command.__call__(self)
        already_exists = False
        if self.funk in self.funk_dict:
            already_exists = True
            log.logger.info(
                'Funk "%s" is already defined. Running edit command.',
                self.funk,
            )

        try:
            self.edit_funk(startinsert=(not already_exists))
            log.logger.info(
                '%s funk "%s".',
                "Edited" if already_exists else "Added",
                self.funk,
            )
        except errors.BlankDefinition as e:
            raise errors.FunkyError(str(e))

        self.commit()
