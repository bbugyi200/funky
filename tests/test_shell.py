# pylint: disable=no-self-use,redefined-outer-name

"""Test --setup-shell and --init options."""

import os
from pathlib import Path
from typing import Iterator, Optional

from _pytest.capture import CaptureFixture
from pytest import fixture, mark
from syrupy.assertion import SnapshotAssertion as Snapshot

from funky import app


params = mark.parametrize

CONFIG_CONTENTS = """\
alias a='echo apple'
alias b='echo berry'
alias c='echo cherry'
"""
SHELL_TO_CONFIG = {"bash": ".bashrc", "zsh": ".zshrc"}


@fixture
def tmp_home(tmp_path: Path) -> Iterator[Path]:
    old_home = os.environ.get("HOME")
    os.environ["HOME"] = str(tmp_path)

    yield tmp_path

    if old_home:
        os.environ["HOME"] = old_home
    else:
        del os.environ["HOME"]


@params("shell", ["bash", "zsh"])
def test_init(shell: str, snapshot: Snapshot, capsys: CaptureFixture) -> None:
    exit_code = app.main(["--init", shell])
    assert exit_code == 0

    captured = capsys.readouterr()
    assert captured.out == snapshot


@params(
    "shell,contents",
    [
        ("bash", None),
        ("bash", CONFIG_CONTENTS),
        ("zsh", None),
        ("zsh", CONFIG_CONTENTS),
    ],
)
def test_setup_shell(
    shell: str, contents: Optional[str], tmp_home: Path, snapshot: Snapshot
) -> None:
    config_file = tmp_home / SHELL_TO_CONFIG[shell]
    if contents:
        config_file.write_text(contents)

    exit_code = app.main(["--setup-shell", shell])
    assert exit_code == 0

    assert config_file.read_text() == snapshot
