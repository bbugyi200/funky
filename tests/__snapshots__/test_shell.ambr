# name: test_init[bash]
  '
  #!/bin/bash
  
  #################################################################################
  #  funky Shell Integration Script
  #
  # ----- WHAT DOES THIS SCRIPT DO?
  # * Source local and global funks on startup.
  # * Source local funks everytime the directory changes.
  # * Define two wrapper functions, `funky` and `gfunky`.
  #
  # ----- INSTALLATION
  # In order to take advantage of the full benefits of funky, this script must be
  # sourced directly into your .bashrc or .zshrc file. Use `funky --ensurepath`
  # to set this up automatically.
  #################################################################################
  
  # The FUNKY_CMD envvar must be overridable (e.g. by test_funky.sh).
  FUNKY_CMD="${FUNKY_CMD:="$(command -v funky)"}"
  
  ##### CREATE FUNKY DATA DIRECTORY
  if [ "$EUID" -eq 0 ]; then
      _home_dir=/root
  else
      _home_dir="$HOME"
  fi
  
  if [[ -n "$XDG_DATA_HOME" ]]; then
      _xdg_data_dir="$XDG_DATA_HOME"/funky
  else
      _xdg_data_dir="$_home_dir"/.local/share/funky
  fi
  
  [[ -d "$_xdg_data_dir" ]] || mkdir -p "$_xdg_data_dir"
  
  ##### HELPER FUNCTIONS
  # Source all global funks.
  _source_globals() { source <(${FUNKY_CMD} --verbose --global); }
  
  # Source local (per-directory) funks.
  _source_locals() { source <(${FUNKY_CMD} --verbose) && _save_locals; }
  
  # Get the current TTY number.
  _tty_number() { tty | sed 's/[^0-9]*//'; }
  
  # Save PWD and associated funks to disk.
  _save_locals() {
      export _ACTIVE_LOCALPATH="$_xdg_data_dir"/localpath-"$(_tty_number)"
      export _ACTIVE_ALIASES="$_xdg_data_dir"/localalias-"$(_tty_number)"
  
      ${FUNKY_CMD} | perl -nE 'print s/\(.+$//gr if /^\S+\(/' >"$_ACTIVE_ALIASES"
  
      if [[ ! -f "$_ACTIVE_LOCALPATH" ]]; then
          echo "$PWD" >"$_ACTIVE_LOCALPATH"
      fi
  }
  
  # Unset funks from last directory.
  _unset_locals() {
      source <(sed 's/^/unset -f /' "${_ACTIVE_ALIASES}")
      command rm -f "$_ACTIVE_ALIASES"
      unset _ACTIVE_ALIASES
  }
  
  # Unset old local funks, if necessary.
  _maybe_unset_locals() {
      if [[ -f "$_ACTIVE_ALIASES" ]]; then
          _unset_locals
      fi
  }
  
  # Activate global funks, if necessary.
  _maybe_source_globals() {
      if [[ -f "${_home_dir}"/.funky ]]; then
          _source_globals
      fi
  }
  
  # Activate local funks, if necessary.
  _maybe_source_locals() {
      if [[ -f "$PWD"/.funky ]] && [[ "$PWD" != "${_home_dir}" ]]; then
          _source_locals
      fi
  }
  
  # Setup the proper funk environment.
  function _setup_funks() {
      _maybe_unset_locals
      _maybe_source_globals
      _maybe_source_locals
  }
  
  ##### SOURCE FUNKS ON STARTUP
  _setup_funks
  
  ##### SOURCE APPROPRIATE FUNKS EVERYTIME THE DIRECTORY IS CHANGED
  #
  # * Is run everytime the directory is changed.
  # * Lazily loads global funks and local funks.
  chpwd() {
      if [[ -f "$_ACTIVE_LOCALPATH" ]] && ! [[ "$PWD" == $(cat "$_ACTIVE_LOCALPATH") ]]; then
          _maybe_unset_locals
          _maybe_source_globals
  
          command rm -f "$_ACTIVE_LOCALPATH"
          unset _ACTIVE_LOCALPATH
      fi
  
      if [[ "$PWD" != "$_home_dir" ]] && [[ -f "$PWD"/.funky ]]; then
          _maybe_source_locals
  
          if [[ ! -f "$_ACTIVE_LOCALPATH" ]]; then
              echo "$PWD" >"$_ACTIVE_LOCALPATH"
          fi
      fi
  }
  
  PROMPT_COMMAND=chpwd
  
  ##### FUNKY'S WRAPPER FUNCTIONS
  # Wrapper used to interact with local funks.
  unalias funky &>/dev/null
  funky() { _funky .funky "$@"; }
  
  # Wrapper used to interact with global funks.
  unalias gfunky &>/dev/null
  gfunky() { _funky ~/.funky --global "$@"; }
  
  # Helper function used by funky's main wrapper functions.
  function _funky() {
      local funk_defs_file="$1"
      shift
  
      touch "$_xdg_data_dir"/timestamp
  
      ${FUNKY_CMD} --color=y "$@"
      if [[ "${funk_defs_file}" -nt "$_xdg_data_dir"/timestamp ]]; then
          _setup_funks
      fi
  }
  
  '
---
# name: test_init[zsh]
  '
  #!/bin/bash
  
  #################################################################################
  #  funky Shell Integration Script
  #
  # ----- WHAT DOES THIS SCRIPT DO?
  # * Source local and global funks on startup.
  # * Source local funks everytime the directory changes.
  # * Define two wrapper functions, `funky` and `gfunky`.
  #
  # ----- INSTALLATION
  # In order to take advantage of the full benefits of funky, this script must be
  # sourced directly into your .bashrc or .zshrc file. Use `funky --ensurepath`
  # to set this up automatically.
  #################################################################################
  
  # The FUNKY_CMD envvar must be overridable (e.g. by test_funky.sh).
  FUNKY_CMD="${FUNKY_CMD:="$(command -v funky)"}"
  
  ##### CREATE FUNKY DATA DIRECTORY
  if [ "$EUID" -eq 0 ]; then
      _home_dir=/root
  else
      _home_dir="$HOME"
  fi
  
  if [[ -n "$XDG_DATA_HOME" ]]; then
      _xdg_data_dir="$XDG_DATA_HOME"/funky
  else
      _xdg_data_dir="$_home_dir"/.local/share/funky
  fi
  
  [[ -d "$_xdg_data_dir" ]] || mkdir -p "$_xdg_data_dir"
  
  ##### HELPER FUNCTIONS
  # Source all global funks.
  _source_globals() { source <(${FUNKY_CMD} --verbose --global); }
  
  # Source local (per-directory) funks.
  _source_locals() { source <(${FUNKY_CMD} --verbose) && _save_locals; }
  
  # Get the current TTY number.
  _tty_number() { tty | sed 's/[^0-9]*//'; }
  
  # Save PWD and associated funks to disk.
  _save_locals() {
      export _ACTIVE_LOCALPATH="$_xdg_data_dir"/localpath-"$(_tty_number)"
      export _ACTIVE_ALIASES="$_xdg_data_dir"/localalias-"$(_tty_number)"
  
      ${FUNKY_CMD} | perl -nE 'print s/\(.+$//gr if /^\S+\(/' >"$_ACTIVE_ALIASES"
  
      if [[ ! -f "$_ACTIVE_LOCALPATH" ]]; then
          echo "$PWD" >"$_ACTIVE_LOCALPATH"
      fi
  }
  
  # Unset funks from last directory.
  _unset_locals() {
      source <(sed 's/^/unset -f /' "${_ACTIVE_ALIASES}")
      command rm -f "$_ACTIVE_ALIASES"
      unset _ACTIVE_ALIASES
  }
  
  # Unset old local funks, if necessary.
  _maybe_unset_locals() {
      if [[ -f "$_ACTIVE_ALIASES" ]]; then
          _unset_locals
      fi
  }
  
  # Activate global funks, if necessary.
  _maybe_source_globals() {
      if [[ -f "${_home_dir}"/.funky ]]; then
          _source_globals
      fi
  }
  
  # Activate local funks, if necessary.
  _maybe_source_locals() {
      if [[ -f "$PWD"/.funky ]] && [[ "$PWD" != "${_home_dir}" ]]; then
          _source_locals
      fi
  }
  
  # Setup the proper funk environment.
  function _setup_funks() {
      _maybe_unset_locals
      _maybe_source_globals
      _maybe_source_locals
  }
  
  ##### SOURCE FUNKS ON STARTUP
  _setup_funks
  
  ##### SOURCE APPROPRIATE FUNKS EVERYTIME THE DIRECTORY IS CHANGED
  #
  # * Is run everytime the directory is changed.
  # * Lazily loads global funks and local funks.
  chpwd() {
      if [[ -f "$_ACTIVE_LOCALPATH" ]] && ! [[ "$PWD" == $(cat "$_ACTIVE_LOCALPATH") ]]; then
          _maybe_unset_locals
          _maybe_source_globals
  
          command rm -f "$_ACTIVE_LOCALPATH"
          unset _ACTIVE_LOCALPATH
      fi
  
      if [[ "$PWD" != "$_home_dir" ]] && [[ -f "$PWD"/.funky ]]; then
          _maybe_source_locals
  
          if [[ ! -f "$_ACTIVE_LOCALPATH" ]]; then
              echo "$PWD" >"$_ACTIVE_LOCALPATH"
          fi
      fi
  }
  
  PROMPT_COMMAND=chpwd
  
  ##### FUNKY'S WRAPPER FUNCTIONS
  # Wrapper used to interact with local funks.
  unalias funky &>/dev/null
  funky() { _funky .funky "$@"; }
  
  # Wrapper used to interact with global funks.
  unalias gfunky &>/dev/null
  gfunky() { _funky ~/.funky --global "$@"; }
  
  # Helper function used by funky's main wrapper functions.
  function _funky() {
      local funk_defs_file="$1"
      shift
  
      touch "$_xdg_data_dir"/timestamp
  
      ${FUNKY_CMD} --color=y "$@"
      if [[ "${funk_defs_file}" -nt "$_xdg_data_dir"/timestamp ]]; then
          _setup_funks
      fi
  }
  
  '
---
# name: test_setup_shell[bash-None]
  '
  
  # setup funky
  command -v funky &>/dev/null && eval "$(funky --init bash)"
  '
---
# name: test_setup_shell[bash-alias a='echo apple'\nalias b='echo berry'\nalias c='echo cherry'\n]
  '
  alias a='echo apple'
  alias b='echo berry'
  alias c='echo cherry'
  
  # setup funky
  command -v funky &>/dev/null && eval "$(funky --init bash)"
  '
---
# name: test_setup_shell[zsh-None]
  '
  
  # setup funky
  command -v funky &>/dev/null && eval "$(funky --init zsh)"
  '
---
# name: test_setup_shell[zsh-alias a='echo apple'\nalias b='echo berry'\nalias c='echo cherry'\n]
  '
  alias a='echo apple'
  alias b='echo berry'
  alias c='echo cherry'
  
  # setup funky
  command -v funky &>/dev/null && eval "$(funky --init zsh)"
  '
---
